security/pkg/server/ca/server.go

```diff
@@ -81,38 +81,18 @@ func (s *Server) CreateCertificate(ctx context.Context, request *pb.IstioCertifi
 	crMetadata := request.Metadata.GetFields()
 	certSigner := crMetadata[security.CertSigner].GetStringValue()
 	log.Debugf("cert signer from workload %s", certSigner)
-	_, _, certChainBytes, rootCertBytes := s.ca.GetCAKeyCertBundle().GetAll()
 	certOpts := ca.CertOpts{
 		SubjectIDs: caller.Identities,
 		TTL:        time.Duration(request.ValidityDuration) * time.Second,
 		ForCA:      false,
 		CertSigner: certSigner,
 	}
-	cert, signErr := s.ca.Sign([]byte(request.Csr), certOpts)
+	respCertChain, signErr := s.ca.SignWithCertChain([]byte(request.Csr), certOpts)
 	if signErr != nil {
 		serverCaLog.Errorf("CSR signing error (%v)", signErr.Error())
 		s.monitoring.GetCertSignError(signErr.(*caerror.Error).ErrorType()).Increment()
 		return nil, status.Errorf(signErr.(*caerror.Error).HTTPErrorCode(), "CSR signing error (%v)", signErr.(*caerror.Error))
 	}
-	respCertChain := []string{string(cert)}
```

Comment:
I don't see why this whole block can be removed. We don't have the same logic in `SignWithCertChain`.

This is changing core code central to the most security critical code in all of Istio, which has been stable for ~3 years. I strongly recommend we do not change this code in order to improve an experimental feature.

If we need a new code path, please make it only for the RA stuff.

`getSignedCertChain` is not sufficient as it does not have the robust error handling and telemetry that we have here

Reply:
My initial suggestion was to not have different code-paths here for the RA and CA. Perhaps we could check if the certSigner was not null and based on that trigger code-paths - for the CA, we could keep the old code, for the RA, the newer code path. @irisdingbj , would this work?

Reply:
Yes.  This works!  Let me keep the ca code path unchanged.